### DISCUSSION TITLE
Flow Diagram

### DISCUSSION THREAD

**Discussion by abhaytiwari262000** [microsoft/presidio#777] [Q&A] [answered]
Hello Omris Sir. I must say that I really like your work and you are an inspiration to me. Also, I have a doubt related to the flow diagram. I want to understand what happens under the hood. when a user uploads an input then what function accepts it and what are the other classes and objects which it calls? (the internal flow diagram )
(PS: I am clear with the concept of regex, context, and Spacy NER. )

**Comment by omri374** [comment, marked as answer]
Hi @abhaytiwari262000, thanks for the kind words. Could you please provide more details on which flow you refer to? We have three main flows in Presidio: text analysis, text anonymization and image analysis+redaction.

**Reply by abhaytiwari262000** [reply]
Sir, text anonymization. In that which is the main class and how does it call other recognizers. Does all recognizers have any base interface ? (So basically, I wanna know the flow )

**Reply by omri374** [reply]
Hi @abhaytiwari262000. The entry point is `AnalyzerEngine`. In the `analyze` method it calls all defined recognizers. [See code here](https://github.com/microsoft/presidio/blob/436938412e8c350618ff6be95f05ba3dc9c0734c/presidio-analyzer/presidio_analyzer/analyzer_engine.py#L171). 
Each recognizer has its own logic for detecting PII, but most of them leverage patterns so they inherit from the `PatternRecognizer` class. 

See this diagram for more information on the structure of recognizers:

![image](https://user-images.githubusercontent.com/3776619/135810537-1a5822b9-2270-498a-b517-48ec72f92cbd.png)


Another part is the NLP analysis, which is done for every request. Here the `NlpEngine` (default is spaCy but the user can select stanza too) creates NLP artifacts (`NlpArtifacts` class) that are passed for each recognizer in case the logic requires tokens, lemmas, named entities or other artifacts created by the `NlpEngine`

See the NLP class structure for more info:

![image](https://user-images.githubusercontent.com/3776619/135810735-e446bbfe-2907-477e-9b24-1343cc61e771.png)

**Reply by abhaytiwari262000** [reply]
Thankyou So Much Sir.
